src/java/org/apache/cassandra/locator/DynamicEndpointSnitch.java

```diff
@@ -23,128 +23,300 @@
+        for (Map.Entry<InetAddressAndPort, AnnotatedMeasurement> entry : samples.entrySet())
         {
-            updateSchedular = ScheduledExecutors.scheduledTasks.scheduleWithFixedDelay(update, dynamicUpdateInterval, dynamicUpdateInterval, TimeUnit.MILLISECONDS);
-            resetSchedular = ScheduledExecutors.scheduledTasks.scheduleWithFixedDelay(reset, dynamicResetInterval, dynamicResetInterval, TimeUnit.MILLISECONDS);
-            registerMBean();
+            // This is expensive for e.g. the Histogram, so do it once and cache the result
+            double measure = entry.getValue().measurement.measure();
+            if (measure > maxLatency)
+                maxLatency = measure;
+            measurements.put(entry.getKey(), measure);
         }
+
+        HashMap<InetAddressAndPort, Double> newScores = new HashMap<>(measurements.size());
+
+        // now make another pass to do the weighting based on the maximums we found before
+        for (Map.Entry<InetAddressAndPort, Double> entry : measurements.entrySet())
+        {
+            double score = entry.getValue() / maxLatency;
+            // finally, add the severity without any weighting, since hosts scale this relative to their own load and the size of the task causing the severity.
+            // "Severity" is basically a measure of compaction activity (CASSANDRA-3722).
+            if (USE_SEVERITY)
+                score += getSeverity(entry.getKey());
+            // lowest score (least amount of badness) wins.
+            newScores.put(entry.getKey(), score);
+        }
+
+        return newScores;
     }
 
     /**
-     * Update configuration from {@link DatabaseDescriptor} and estart the update-scheduler and reset-scheduler tasks
+     * Update configuration from {@link DatabaseDescriptor} and restart the various scheduler tasks
      * if the configured rates for these tasks have changed.
      */
-    public void applyConfigChanges()
+    public void applyConfigChanges(int newDynamicUpdateInternal, int newDynamicLatencyProbeInterval, double newDynamicBadnessThreshold)
     {
-        if (dynamicUpdateInterval != DatabaseDescriptor.getDynamicUpdateInterval())
+        if (dynamicUpdateInterval != newDynamicUpdateInternal)
         {
-            dynamicUpdateInterval = DatabaseDescriptor.getDynamicUpdateInterval();
+            dynamicUpdateInterval = newDynamicUpdateInternal;
             if (DatabaseDescriptor.isDaemonInitialized())
             {
-                updateSchedular.cancel(false);
-                updateSchedular = ScheduledExecutors.scheduledTasks.scheduleWithFixedDelay(update, dynamicUpdateInterval, dynamicUpdateInterval, TimeUnit.MILLISECONDS);
+                if (updateScoresScheduler != null)
+                    updateScoresScheduler.cancel(false);
+                updateScoresScheduler = ScheduledExecutors.scheduledTasks.scheduleWithFixedDelay(this::updateScores, dynamicUpdateInterval, dynamicUpdateInterval, TimeUnit.MILLISECONDS);
```

Comment:
Should this be more "maybeScheduleProbes"

Reply:
Yea ... I can move it onto a dedicated STPE but I don't want to change too much behavior. Regarding the naming the LegacyHistogram doesn't send any probes, so I thought it could just be generically a `updateSamples` method similar to `updateScores` ... I don't care strongly one way or the other it's up to you :-)